### PR TITLE
Example to create ROSA HCP Cluster

### PR DESCRIPTION
**Pull request by tphee**
### Added:
cluster_rosa_hcp_resource.go          # Create ROSA HCP cluster
cluster_rosa_hcp_state.go                 # Create ROSA HCP cluster
cluster_hcp_waiter_resource.go       # Wait for cluster current_compute == desired_compute
cluster_hcp_waiter_state.go              # Wait for cluster current_compute == desired_compute
cluster_data_source.go                      # Retrive cluster api_url and console_url
cluster_data_state.go                         # Retrive cluster api_url and console_url
examples/create_rosa_hcp_cluster/ # Example to create ROSA HCP cluster

### Modified:
ocm_policies_data_source.go             # Support HCP policies
ocm_policies_state.go                          # Support HCP policies
rosa_operator_roles_data_source.go # Support HCP operator roles
rosa_operator_roles_state.go              # Support HCP operator roles
provider.go                                              # Support HCP

### Example to create ROSA HCP Cluster:
Prerequisites: You have created an AWS VPC with public and private subnets.
```
make install
cd examples/create_rosa_hcp_cluster/hcp_cluster_with_managed_oidc_config
export TF_VAR_account_role_prefix=<account_role_prefix>
export TF_VAR_operator_role_prefix=<operator_role_prefix>
export TF_VAR_availability_zones='["<aws_az_region>"]'
export TF_VAR_aws_subnet_ids='["<aws_public_subnetid>","<aws_private_subnetid>"]'
export TF_VAR_cloud_region=<aws_region_name>
export TF_VAR_token=<ocm_offline_token>
terraform init
terraform apply
```

### Example terraform apply output:
```
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.create_account_roles.module.rosa_account_roles[0].random_string.default_random: Creating...
random_password.password: Creating...
module.create_account_roles.module.rosa_account_roles[0].random_string.default_random: Creation complete after 0s [id=k7xv]
random_password.password: Creation complete after 0s [id=none]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[6].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[0].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[4].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[3].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[1].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[1].aws_iam_role.account_role: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[0].aws_iam_policy.account_role_policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[0].aws_iam_role.account_role: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[6].aws_iam_policy.operator-policy: Creation complete after 0s [id=arn:aws:iam::XXXXXXXXXX:policy/rosa-hcp-kube-system-kms-provider]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[2].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[4].aws_iam_policy.operator-policy: Creation complete after 0s [id=arn:aws:iam::XXXXXXXXXX:policy/rosa-hcp-kube-system-capa-controller-manager]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[0].aws_iam_policy.account_role_policy: Creation complete after 0s [id=arn:aws:iam::XXXXXXXXXX:policy/rosa-hcp-Installer-Role-Policy]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[7].aws_iam_policy.operator-policy: Creation complete after 0s [id=arn:aws:iam::XXXXXXXXXX:policy/rosa-hcp-kube-system-kube-controller-manager]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_operator_role_policies[5].aws_iam_policy.operator-policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[1].aws_iam_policy.account_role_policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[1].aws_iam_policy.instance_account_role_policy: Creating...
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[0].aws_iam_role.account_role: Creation complete after 0s [id=rosa-hcp-Installer-Role]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[1].aws_iam_role.instance_account_role: Creation complete after 0s [id=rosa-hcp-ControlPlane-Role]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[0].aws_iam_policy_attachment.instance_role_policy_attachment: Creation complete after 0s [id=worker-role-policy-attachment]
module.create_account_roles.module.rosa_account_roles[0].module.rosa_account_roles[1].aws_iam_policy_attachment.role_policy_attachment: Creation complete after 0s [id=support-role-policy-attachment]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[4].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].data.aws_caller_identity.current: Reading...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[4].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].data.aws_caller_identity.current: Read complete after 0s [id=XXXXXXXXXX]
rhcs_rosa_oidc_config.oidc_config: Creation complete after 3s [id=2634i25cu7i6pdnmb2l3a4k3pg7csd52]
module.operator_roles_and_oidc_provider.module.rosa_oidc_provider[0].aws_iam_openid_connect_provider.oidc_provider: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[4].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].aws_iam_role.operator_role: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-openshift-cluster-csi-drivers-ebs-cloud-credentials]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-kube-system-kms-provider]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[4].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-kube-system-control-plane-operator]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-openshift-ingress-operator-cloud-credentials]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-openshift-image-registry-installer-cloud-credentials]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-openshift-cloud-network-config-controller-cloud-credentia]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-kube-system-kube-controller-manager]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[4].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].aws_iam_role.operator_role: Creation complete after 0s [id=rosa-hcp-kube-system-capa-controller-manager]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creating...
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[6].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-kube-system-kms-provider-20230906151314474200000002]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[1].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-kube-system-capa-controller-manager-20230906151314491700000004]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[0].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-openshift-cluster-csi-drivers-ebs-cloud-credentials-20230906151314481400000003]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[5].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-openshift-ingress-operator-cloud-credentials-20230906151314492400000005]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[2].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-openshift-cloud-network-config-controller-cloud-credentia-20230906151314498300000007]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[7].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-kube-system-kube-controller-manager-20230906151314498600000008]
module.operator_roles_and_oidc_provider.module.rosa_operator_roles[0].module.rosa_operator_roles[3].aws_iam_role_policy_attachment.operator_role_policy_attachment: Creation complete after 1s [id=rosa-hcp-openshift-image-registry-installer-cloud-credentials-20230906151314498300000006]
time_sleep.wait_10_seconds: Creating...
time_sleep.wait_10_seconds: Creation complete after 10s [id=2023-09-06T15:13:24Z]
time_sleep.wait_10_seconds: Still creating... [10s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Creating...
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [10s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [20s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [30s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [40s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [50s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [1m0s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [1m10s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [1m20s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [9m20s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [9m30s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Still creating... [9m40s elapsed]
rhcs_cluster_rosa_hcp.rosa_hcp_cluster: Creation complete after 9m41s [id=2634i53i41uapt9s7avhbndteesqvmd4]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Creating...
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [10s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [20s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [30s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [40s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [50s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [1m0s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [1m10s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m1s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m11s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m21s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m31s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m41s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [15m51s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Still creating... [16m1s elapsed]
rhcs_cluster_hcp_wait.rosa_cluster_hcp: Creation complete after 16m2s
data.rhcs_cluster_data.cluster: Reading...
data.rhcs_cluster_data.cluster: Read complete after 0s

Apply complete! Resources: 43 added, 0 changed, 0 destroyed.

Outputs:

account_role_prefix = "rosa-hcp"
api_url = "https://api.rosa-hcp.elgb.p3.openshiftapps.com:443"
cloud_region = "us-west-2"
cluster_admin_username = "cluster-admin"
cluster_id = "2634i53i41uapt9s7avhbndteesqvmd4"
cluster_name = "rosa-hcp"
console_url = "https://console-openshift-console.apps.rosa.rosa-hcp.elgb.p3.openshiftapps.com"
openshift_version = "4.13.8"
```

### CODE REVIEW

File: provider/cluster_data_source.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
```
Comment (bardielle) [inline provider/cluster_data_source.go:24]: We upgraded the framework so the interface looks different 
Please rebace on top of the main branch
Also, I'm working on cluster data source to support more attributes and soon will open a PR for that

File: examples/create_rosa_hcp_cluster/hcp_cluster_with_managed_oidc_config/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# Creating a ROSA HCP cluster with a managed OIDC configuration ID
```
Comment (bardielle) [inline examples/create_rosa_hcp_cluster/hcp_cluster_with_managed_oidc_config/README.md:1]: I prefer you will move that example dir into the new [module](https://github.com/terraform-redhat/terraform-rhcs-rosa)

File: provider/cluster_hcp_waiter_resource.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	hcpDefaultTimeoutInMinutes   = int64(60)
+	hcpNonPositiveTimeoutSummary = "Can't poll cluster state with a non-positive timeout"
+	hcpNonPositiveTimeoutFormat  = "Can't poll state of cluster with identifier '%s', the timeout that was set is not a positive number"
+	hclPollingIntervalInMinutes  = 2
+)
+
+func (t *ClusterHCPWaiterResourceType) GetSchema(ctx context.Context) (result tfsdk.Schema,
+	diags diag.Diagnostics) {
+	result = tfsdk.Schema{
+		Description: "Wait Cluster Resource To be Ready when current compute is equals to desired compute",
```
Comment (bardielle) [inline provider/cluster_hcp_waiter_resource.go:37]: Why can't it be part of cluster_rosa_hcp resource?

File: provider/cluster_rosa_hcp_resource.go [hunk 4]
```diff
@@ -0,0 +1,2036 @@
...
+					},
+					"additional_trust_bundle": {
+						Description: "A string containing a PEM-encoded X.509 certificate bundle that will be added to the nodes' trusted certificate store.",
+						Type:        types.StringType,
+						Optional:    true,
+					},
+				}),
+				Optional:   true,
+				Validators: proxyValidators(),
+			},
```
Comment (bardielle) [inline provider/cluster_rosa_hcp_resource.go:324]:
```suggestion
"proxy": schema.SingleNestedAttribute{
				Description: "proxy",
				Attributes:  proxy.ProxyResource(),
				Optional:    true,
				Validators:  []validator.Object{proxy.ProxyValidator()},
			},
```

File: provider/cluster_rosa_hcp_state.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+*/
+
+package provider
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type ClusterRosaHCPState struct {
+	APIURL                    types.String      `tfsdk:"api_url"`
```
Comment (bardielle) [inline provider/cluster_rosa_hcp_state.go:24]: please add a comment with the new/different attributes from the cluster_rosa_Classic resource

File: provider/cluster_data_source.go (hunk 1 above)
Comment (tphee) [reply provider/cluster_data_source.go:24]: The framework has changed so much.  I would have to re-fork the repository and re-do the work.

**Comment by tphee** [comment]
Closing this.  Will open a new PR.

**Pull request closed without merging**